src/components/fast_clock.c

```diff
@@ -0,0 +1,253 @@
+  NVIC_EnableIRQ(nvic_irqn);
+
+  fast_clock_timer->COUNT16.COUNT.reg = 0;
+  fast_clock_timer->COUNT16.CC[0].reg = FAST_CLOCK_TIMER_DELAY;
+  fast_clock_timer->COUNT16.CTRLA.reg |= TC_CTRLA_ENABLE;
+}
+
+// ----------------------------------------------------------------------------
+// sets the values for being a master: whether it is a
+void fast_clock_set_as_master(uint8_t id1, uint8_t id2, uint16_t intermessage_delay)
+{
+  fast_clock_status.master = true;
+  fast_clock_status.id1 = id1;
+  fast_clock_status.id2 = id2;
+  fast_clock_status.intermessage_delay = intermessage_delay;
+}
+
+// ----------------------------------------------------------------------------
+// sets the thing as a slave.
+void fast_clock_set_as_slave(void)
+{
+  fast_clock_status.master = false;
+}
+
+//----------------------------------------------------------------------------
+void fast_clock_set_time(FAST_CLOCK_TIME_Type time)
+{
+  // Set the time
+  current_time = time;
+  // Reset the millisecond counter
+  fast_clock_status.millisecond = 0;
+  // Notify the update!
+  fast_clock_handle_update(current_time);
+}
+
+//----------------------------------------------------------------------------
+// this function sets the clock rate.
+void fast_clock_set_rate(uint8_t rate)
+{
+  fast_clock_status.rate = rate;
+}
+
+//----------------------------------------------------------------------------
+// this function processes the clock information sent on the loconet
+void loconet_rx_fast_clock(uint8_t *data, uint8_t length)
+{
+  if (length < 8)
+  {
+    return;
+  }
+
+  if (!(data[7] == 1)) {
+    // the message is not a correct clock tick!
+    return;
+  }
+
+  //1st byte of data is the clock rate
+  fast_clock_set_rate(data[0]);
+
+  current_time.minute = data[3] - (128-60);
+  current_time.hour = data[5] >= (128-24) ? data[5] - (128-24) : data[5] % 24;
+  current_time.day = data[6] % 7;
+  current_time.second = 0;
+
+  fast_clock_status.millisecond = 0;
+
+  fast_clock_handle_update(current_time);
+}
+
+//-----------------------------------------------------------------------------
+// sends the message in the appropriate format
+static void fast_clock_send_message(void)
+{
+  loconet_tx_fast_clock(
+    fast_clock_status.rate,
+    0x0,
+    0x0,
+    current_time.minute,
+    current_time.hour,
+    current_time.day,
+    fast_clock_status.id1,
+    fast_clock_status.id2
+  );
+}
+
+
+// ----------------------------------------------------------------------------
+// this function should be called every 50ms. It updates the clock with
+// the set rate.
+void fast_clock_tick(void)
+{
+  // keeps track whether we need an update. We only send an update
+  // if the minute counter has been increased.
+  bool notify = false;
+
+  // update the milliseconds passed with the clock rate
+  fast_clock_status.millisecond += fast_clock_status.rate;
+
+  // as the millisecond counter counts every 50ms, we have 200 cycles for a
```

Comment:
Explained in comments. In fact I should add 50 * rate to the millisec counter, and then check whether this is >= 1000. Just dividing the lot by 50 is simpler :)